pyquil/api/_quantum_computer.py

```diff
@@ -129,28 +129,26 @@ def to_compiler_isa(self) -> CompilerISA:
         """
         return self.compiler.quantum_processor.to_compiler_isa()
 
-    @_record_call
     def run(
         self,
         executable: QuantumExecutable,
-        memory_map: Optional[Mapping[str, Sequence[Union[int, float]]]] = None,
     ) -> np.ndarray:
         """
-        Run a quil executable. If the executable contains declared parameters, then a memory
-        map must be provided, which defines the runtime values of these parameters.
+        Run a quil executable. All parameters in the executable must have values applied using
+        ``Program#with_parameter_values`` or ``Program#set_parameter_value``.
 
-        :param executable: The program to run. You are responsible for compiling this first.
-        :param memory_map: The mapping of declared parameters to their values. The values
-            are a list of floats or integers.
+        :param executable: The program to run, compiled as needed for its target QAM.
         :return: A numpy array of shape (trials, len(ro-register)) that contains 0s and 1s.
         """
-        self.qam.load(executable)
-        if memory_map:
-            for region_name, values_list in memory_map.items():
-                self.qam.write_memory(region_name=region_name, value=values_list)
-        return self.qam.run().wait().read_memory(region_name="ro")  # type: ignore
+        result = self.qam.run(executable)
+
+        # QUESTION (for @ameyer) - this is consistent with the V2 API, but in that API,
+        # QuantumComputer#run and QAM#run return different shapes.
+        #
+        # Does it make sense to align those now and return a QAMExecutionResult, given that
+        # it's a breaking change, or is it best to leave this as-is for ease/convenience?
+        return result.read_memory(region_name="ro")
```

Comment:
I see what you mean. We should make them `QAM.run()` and `QuantumComputer.run()` consistent with each other in this new API. It also has the nice benefit of removing assumptions about wanting to read `ro` (we've already removed implicit `ro` declarations, so this would make sense).

Reply:
Then, since `QuantumComputer.run` is just a one-line delegation to `self.qam.run()`... should we just remove `QuantumComputer.run()` altogether?

Reply:
Meh, I'd say keep it because it seems intuitive to be able to run directly via `QuantumComputer`. Just my opinion though.

Reply:
I'll keep it, but now it will return a `QAMExecutionResult`. lmk if that checks out

Reply:
Yeah, that makes sense. I like the consistency of operating off of `QAMExecutionResult`s in the end, whether users call `execute` or `run`. The only difference between behavior then becomes async vs sync, which is the important difference (tossing in a result shape change might be confusing).